# BATTLE OF NEIGHBORHOOD -- Report (Week 1 and 2)

## Business Problem 

#### Background

Finding a suitable property in the urban city is a very difficult job. One does not have enough time to invest in searching property plus sometimes real estate agents also hide information from the client which later become issue for the client. 
There are problems such as: hidden price falls, record-low sales, homebuilder exodus and tax hikes addressing overseas buyers of homes in Brooklyn.

Person who is looking for buying a new property for investment purpose will get benefit with this project.

#### Business Problem

How could we provide support to homebuyers clientele in to purchase a suitable real estate in Brooklyn keeping clients basic needs in mind?

To solve this problem we will make clusters of the property based on the basic amenities and essential facilities surrounding such venues i.e. elementary schools, high schools, supermarket & restaurants.

## Data Collection

Data on Brooklyn property and its price were extracted from the New York Goverment site (http://www1.nyc.gov). The attributes of the file contain Borough, Neighborhood, Building class category, Address, Price, Year built.

Link: https://www1.nyc.gov/assets/finance/downloads/pdf/rolling_sales/rollingsales_brooklyn.xlx

## Methodlogy

The Methodology section comprises four stages:

1. Explore data
2. Data preparation and preprocessing 
3. Modeling
4. Analyzing Cluster

### 1. Explore  Data

In [4]:
import numpy as np # library to handle data in a vectorized manner

import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json # library to handle JSON files
import locale
#!conda install -c conda-forge geopy --yes # uncomment this line if you haven't completed the Foursquare API lab
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans
import datetime as dt

#!conda install -c conda-forge folium=0.5.0 --yes # uncomment this line if you haven't completed the Foursquare API lab
import folium # map rendering library

print('Libraries imported.')

Libraries imported.


#### Data Loading and Cleaning 

In this section, Property Sales data for the Brooklyn area is loaded and cleaned. The following data contain various attributes such as Sale price, Age of the building, Building Category etc.

In [5]:
proj_data=pd.read_excel('rollingsales_brooklyn.xls')
proj_data.head()

BOROUGH NEIGHBORHOOD  BUILDING CLASS CATEGORY   BLOCK   LOT  \
0      3.0   BATH BEACH  01 ONE FAMILY DWELLINGS  6360.0  23.0   
1      3.0   BATH BEACH  01 ONE FAMILY DWELLINGS  6363.0  48.0   
2      3.0   BATH BEACH  01 ONE FAMILY DWELLINGS  6363.0  48.0   
3      3.0   BATH BEACH  01 ONE FAMILY DWELLINGS  6366.0  69.0   
4      3.0   BATH BEACH  01 ONE FAMILY DWELLINGS  6366.0  72.0   

                ADDRESS  ZIP CODE  RESIDENTIAL UNITS  COMMERCIAL UNITS  \
0      8645 15TH AVENUE   11228.0                1.0               0.0   
1    12 BAY 13TH STREET   11214.0                1.0               0.0   
2  8658 BAY 16TH STREET   11214.0                1.0               0.0   
3  8664 BAY 16TH STREET   11214.0                1.0               0.0   
4      1728 86TH STREET   11214.0                1.0               0.0   

   TOTAL UNITS LAND SQUARE FEET  GROSS SQUARE FEET  YEAR BUILT  SALE PRICE  \
0          1.0             1547             1428.0      1930.0    750000.0   
1          1.0             3142             3200.0      1999.0         0.0   
2          1.0             3142             3200.0      1999.0         0.0   
3          1.0             4833             1724.0      1930.0         0.0   
4          1.0             4833             2300.0      1925.0   1720000.0   

   SALE DATE  
0 2018-05-18  
1 2019-02-11  
2 2019-02-27  
3 2018-10-25  
4 2018-12-12

### 2. Data preparation and preprocessing

In [6]:
proj_data.dropna(subset=['ADDRESS','SALE DATE'],inplace=True,axis=0)
proj_data['Total Years']=proj_data['SALE DATE'].dt.year.astype('int')-proj_data['YEAR BUILT']
proj_data.drop(inplace=True,columns=['SALE DATE','YEAR BUILT','GROSS SQUARE FEET'])
proj_data=proj_data[proj_data['SALE PRICE']!=0.0]
proj_data['BOROUGH']='Brooklyn'
proj_data.reset_index(drop=True).head(10)

BOROUGH NEIGHBORHOOD  BUILDING CLASS CATEGORY   BLOCK   LOT  \
0  Brooklyn   BATH BEACH  01 ONE FAMILY DWELLINGS  6360.0  23.0   
1  Brooklyn   BATH BEACH  01 ONE FAMILY DWELLINGS  6366.0  72.0   
2  Brooklyn   BATH BEACH  01 ONE FAMILY DWELLINGS  6367.0  41.0   
3  Brooklyn   BATH BEACH  01 ONE FAMILY DWELLINGS  6367.0  42.0   
4  Brooklyn   BATH BEACH  01 ONE FAMILY DWELLINGS  6367.0  43.0   
5  Brooklyn   BATH BEACH  01 ONE FAMILY DWELLINGS  6380.0  73.0   
6  Brooklyn   BATH BEACH  01 ONE FAMILY DWELLINGS  6397.0  63.0   
7  Brooklyn   BATH BEACH  01 ONE FAMILY DWELLINGS  6401.0  10.0   
8  Brooklyn   BATH BEACH  01 ONE FAMILY DWELLINGS  6402.0   4.0   
9  Brooklyn   BATH BEACH  01 ONE FAMILY DWELLINGS  6460.0  59.0   

               ADDRESS  ZIP CODE  RESIDENTIAL UNITS  COMMERCIAL UNITS  \
0     8645 15TH AVENUE   11228.0                1.0               0.0   
1     1728 86TH STREET   11214.0                1.0               0.0   
2     1730 86TH STREET   11214.0                1.0               1.0   
3     1732 86TH STREET   11214.0                1.0               1.0   
4     8672 BAY PARKWAY   11214.0                1.0               1.0   
5    132 BAY 13 STREET   11214.0                1.0               1.0   
6     1725 BATH AVENUE   11214.0                1.0               0.0   
7     1763 BATH AVENUE   11214.0                1.0               1.0   
8   1906 BENSON AVENUE   11214.0                1.0               1.0   
9  286 BAY 10TH STREET   11228.0                1.0               0.0   

   TOTAL UNITS LAND SQUARE FEET  SALE PRICE  Total Years  
0          1.0             1547    750000.0         88.0  
1          1.0             4833   1720000.0         93.0  
2          2.0             1342   1380000.0         87.0  
3          2.0             1342   1630000.0         93.0  
4          2.0             1342   1630000.0         93.0  
5          2.0             1740    968000.0         58.0  
6          1.0             2708   1150000.0        109.0  
7          2.0             1160    980000.0         99.0  
8          2.0              980        10.0         91.0  
9          1.0             3867   1660000.0         93.0

Following data shows total property grouped by Neighborhood of the Brooklyn.

In [7]:
proj_data.groupby(by=['NEIGHBORHOOD']).count()

BOROUGH  BUILDING CLASS CATEGORY  BLOCK  LOT  \
NEIGHBORHOOD                                                             
BATH BEACH                    208                      208    208  208   
BAY RIDGE                     565                      565    565  565   
BEDFORD STUYVESANT            853                      853    853  853   
BENSONHURST                   284                      284    284  284   
BERGEN BEACH                  125                      125    125  125   
BOERUM HILL                   152                      152    152  152   
BOROUGH PARK                  545                      545    545  545   
BRIGHTON BEACH                232                      232    232  232   
BROOKLYN HEIGHTS              291                      291    291  291   
BROWNSVILLE                   126                      126    126  126   
BUSH TERMINAL                  24                       24     24   24   
BUSHWICK                      420                      420    420  420   
CANARSIE                      548                      548    548  548   
CARROLL GARDENS               121                      121    121  121   
CLINTON HILL                  235                      235    235  235   
COBBLE HILL                    97                       97     97   97   
COBBLE HILL-WEST               64                       64     64   64   
CONEY ISLAND                  201                      201    201  201   
CROWN HEIGHTS                 478                      478    478  478   
CYPRESS HILLS                 214                      214    214  214   
DOWNTOWN-FULTON FERRY          97                       97     97   97   
DOWNTOWN-FULTON MALL          108                      108    108  108   
DOWNTOWN-METROTECH             94                       94     94   94   
DYKER HEIGHTS                 168                      168    168  168   
EAST NEW YORK                 676                      676    676  676   
FLATBUSH-CENTRAL              395                      395    395  395   
FLATBUSH-EAST                 566                      566    566  566   
FLATBUSH-LEFFERTS GARDEN      114                      114    114  114   
FLATBUSH-NORTH                255                      255    255  255   
FLATLANDS                     161                      161    161  161   
FORT GREENE                   160                      160    160  160   
GERRITSEN BEACH               108                      108    108  108   
GOWANUS                       118                      118    118  118   
GRAVESEND                     448                      448    448  448   
GREENPOINT                    319                      319    319  319   
JAMAICA BAY                     1                        1      1    1   
KENSINGTON                    101                      101    101  101   
MADISON                       353                      353    353  353   
MANHATTAN BEACH                54                       54     54   54   
MARINE PARK                   256                      256    256  256   
MIDWOOD                       293                      293    293  293   
MILL BASIN                     72                       72     72   72   
NAVY YARD                      18                       18     18   18   
OCEAN HILL                    271                      271    271  271   
OCEAN PARKWAY-NORTH           385                      385    385  385   
OCEAN PARKWAY-SOUTH           174                      174    174  174   
OLD MILL BASIN                178                      178    178  178   
PARK SLOPE                    531                      531    531  531   
PARK SLOPE SOUTH              147                      147    147  147   
PROSPECT HEIGHTS              257                      257    257  257   
RED HOOK                       37                       37     37   37   
SEAGATE                        33                       33     33   33   
SHEEPSHEAD BAY                

#### We are taking Brooklyn Heights as a Neighborhood for this project.

In [8]:
brook_height=proj_data[proj_data['NEIGHBORHOOD']=='BROOKLYN HEIGHTS']
brook_height.head()

BOROUGH      NEIGHBORHOOD  BUILDING CLASS CATEGORY  BLOCK    LOT  \
4927  Brooklyn  BROOKLYN HEIGHTS  01 ONE FAMILY DWELLINGS  208.0  316.0   
4929  Brooklyn  BROOKLYN HEIGHTS  01 ONE FAMILY DWELLINGS  215.0   19.0   
4931  Brooklyn  BROOKLYN HEIGHTS  01 ONE FAMILY DWELLINGS  220.0    5.0   
4932  Brooklyn  BROOKLYN HEIGHTS  01 ONE FAMILY DWELLINGS  225.0   10.0   
4933  Brooklyn  BROOKLYN HEIGHTS  01 ONE FAMILY DWELLINGS  252.0   42.0   

                  ADDRESS  ZIP CODE  RESIDENTIAL UNITS  COMMERCIAL UNITS  \
4927  1121 MADISON STREET   11201.0                1.0               0.0   
4929  1095 MADISON STREET   11201.0                1.0               0.0   
4931  1249 MADISON STREET   11201.0                1.0               0.0   
4932   460 CENTRAL AVENUE   11201.0                1.0               0.0   
4933   1245 PUTNAM AVENUE   11201.0                1.0               0.0   

      TOTAL UNITS LAND SQUARE FEET  SALE PRICE  Total Years  
4927          1.0             2492  11750000.0        162.0  
4929          1.0             1358   3600000.0        119.0  
4931          1.0             1263   2530000.0        120.0  
4932          1.0             1017   4350000.0        165.0  
4933          1.0             1449   5000000.0        108.0

#### Since we don't know the latitude and longitude of the buildings, in the following code we fetch all the coordinates of the building by using GeoCoder.
Limitation: If we are iterating a large set of data then after few minutes geocoder will deny your request which leads to error '429 TOO MANY REQUESTS'. 
This is the main reason for choosing only one neighborhood.

In [9]:
latitude=[]
longitude=[]
i=0
for addr,neigh in zip(brook_height['ADDRESS'],brook_height['NEIGHBORHOOD']):
    address = addr+' ,'+neigh+ ' ,Brooklyn, NY'
    #print(address,'-',i)
    geolocator = Nominatim(user_agent="ny_explorer",timeout=3)
    location = geolocator.geocode(address)
    """geocoder = GoogleGeocoder('AIzaSyBk7EELr6hKx3h4E_5PDvNnCKDRmtomwbc')
    search = geocoder.get(address)"""
    try:
        latitude.append(location.latitude)
        longitude.append(location.longitude)
    except:
        latitude.append(np.nan)
        longitude.append(np.nan)
    i+=1
    if(i==50):
        print(latitude)
brook_height['latitude']=latitude
brook_height['longitude']=longitude
brook_height.head()

[40.6989204, 40.6989204, 40.6989204, nan, 40.682587, 40.682587, 40.6989204, nan, 40.682587, 40.6978252, 40.682587, 40.682587, 40.6978252, 40.69584275, 40.6842279, 40.6846267, 40.6842279, 40.6846267, nan, nan, 40.68992665, 40.68967915, 40.6913148, 40.6949636, 40.6857874285714, 40.6886903, 40.68384, 40.68384, 40.692549, 40.68384, 40.6955489591837, 40.695540122449, 40.692549, 40.692549, 40.6945351, nan, 40.6954738469388, 40.6955091938776, 40.6955180306122, 40.6955622142857, 40.6945351, 40.6945351, nan, nan, 40.6945351, 40.6945351, 40.6945351, nan, nan, 40.6945351]


/home/jupyterlab/conda/lib/python3.6/site-packages/ipykernel_launcher.py:20: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
/home/jupyterlab/conda/lib/python3.6/site-packages/ipykernel_launcher.py:21: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy


BOROUGH      NEIGHBORHOOD  BUILDING CLASS CATEGORY  BLOCK    LOT  \
4927  Brooklyn  BROOKLYN HEIGHTS  01 ONE FAMILY DWELLINGS  208.0  316.0   
4929  Brooklyn  BROOKLYN HEIGHTS  01 ONE FAMILY DWELLINGS  215.0   19.0   
4931  Brooklyn  BROOKLYN HEIGHTS  01 ONE FAMILY DWELLINGS  220.0    5.0   
4932  Brooklyn  BROOKLYN HEIGHTS  01 ONE FAMILY DWELLINGS  225.0   10.0   
4933  Brooklyn  BROOKLYN HEIGHTS  01 ONE FAMILY DWELLINGS  252.0   42.0   

                  ADDRESS  ZIP CODE  RESIDENTIAL UNITS  COMMERCIAL UNITS  \
4927  1121 MADISON STREET   11201.0                1.0               0.0   
4929  1095 MADISON STREET   11201.0                1.0               0.0   
4931  1249 MADISON STREET   11201.0                1.0               0.0   
4932   460 CENTRAL AVENUE   11201.0                1.0               0.0   
4933   1245 PUTNAM AVENUE   11201.0                1.0               0.0   

      TOTAL UNITS LAND SQUARE FEET  SALE PRICE  Total Years   latitude  \
4927          1.0             2492  11750000.0        162.0  40.698920   
4929          1.0             1358   3600000.0        119.0  40.698920   
4931          1.0             1263   2530000.0        120.0  40.698920   
4932          1.0             1017   4350000.0        165.0        NaN   
4933          1.0             1449   5000000.0        108.0  40.682587   

      longitude  
4927 -73.909293  
4929 -73.909293  
4931 -73.909293  
4932        NaN  
4933 -73.962611

In [10]:
brook_prop=brook_height[['BUILDING CLASS CATEGORY','ADDRESS','SALE PRICE','Total Years','latitude','longitude']].copy()
brook_prop.head()

BUILDING CLASS CATEGORY              ADDRESS  SALE PRICE  Total Years  \
4927  01 ONE FAMILY DWELLINGS  1121 MADISON STREET  11750000.0        162.0   
4929  01 ONE FAMILY DWELLINGS  1095 MADISON STREET   3600000.0        119.0   
4931  01 ONE FAMILY DWELLINGS  1249 MADISON STREET   2530000.0        120.0   
4932  01 ONE FAMILY DWELLINGS   460 CENTRAL AVENUE   4350000.0        165.0   
4933  01 ONE FAMILY DWELLINGS   1245 PUTNAM AVENUE   5000000.0        108.0   

       latitude  longitude  
4927  40.698920 -73.909293  
4929  40.698920 -73.909293  
4931  40.698920 -73.909293  
4932        NaN        NaN  
4933  40.682587 -73.962611

In [11]:
brook_prop.dropna(subset=['latitude'],inplace=True,axis=0)

### After retreiving the coordinates of the buildings I plotted them on the map. 
Note: Few building have the same coordinates but the address is different. 

In [12]:
address = 'Brooklyn, New York City, NY'

geolocator = Nominatim(user_agent="ny_explorer")
location = geolocator.geocode(address)
latitudes = location.latitude
longitudes = location.longitude

print('The geograpical coordinate of Brooklyn, New York City is {}, {}.'.format(latitudes, longitudes))

The geograpical coordinate of Brooklyn, New York City is 40.6501038, -73.9495823.


In [84]:
map_newyork = folium.Map(location=[latitudes, longitudes], zoom_start=10)
for lat, lng, BC_cat, address,price,total_yr in zip(brook_prop['latitude'], brook_prop['longitude'], 
                                                   brook_prop['BUILDING CLASS CATEGORY'], brook_prop['ADDRESS'],
                                                   brook_prop['SALE PRICE'],brook_prop['Total Years']):
    label = 'Building Cat: {},<br>Address: {},<br>Price: {},<br>Total Year: {}'.format(BC_cat,address, price ,total_yr)
    label = folium.Popup(label)
    folium.CircleMarker(
        [lat, lng],
        radius=4,
        popup=label,
        color='black',
        fill=True,
        fill_color='black',
        fill_opacity=0.7,
        parse_html=False).add_to(map_newyork)

 
# Save to html
map_newyork.save('Plot_map.html')

#### Analyzing Surroundings using FOURSQUARE

In [14]:
CLIENT_ID = 'QKV5PUM5COLHDT35BNNHD4JWCQLHRUWHAR43ESK3OQULVEQD' # your Foursquare ID
CLIENT_SECRET = 'ZZK3GYORWJFC3IDVLW4NKI0EWOZGZ12BHNS1V2WHJA40UBFT' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: QKV5PUM5COLHDT35BNNHD4JWCQLHRUWHAR43ESK3OQULVEQD
CLIENT_SECRET:ZZK3GYORWJFC3IDVLW4NKI0EWOZGZ12BHNS1V2WHJA40UBFT


#### Defining a function for getting all the venues near the building within radius of 1 KM

In [15]:
def getNearbyVenues(names, address, latitudes, longitudes, radius=1000):
    LIMIT=150
    venues_list=[]
    for name, address, lat, lng in zip(names, address,latitudes, longitudes):
        #print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name,
            address,
            lat, 
            lng, 
            v['venue']['name'],
            v['venue']['location']['distance'],
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Building Class Name','Address', 
                  'Building Class Latitude', 
                  'Building Class Longitude', 
                  'Venue',
                  'Distance',
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [16]:
brooklyn_venues = getNearbyVenues(names=brook_prop['BUILDING CLASS CATEGORY'],
                                   address=brook_prop['ADDRESS'],
                                   latitudes=brook_prop['latitude'],
                                   longitudes=brook_prop['longitude']
                                  )
brooklyn_venues.shape

(17636, 9)

As you can see we got 17636 venues in total.

In [17]:
brooklyn_venues.head()

Building Class Name              Address  Building Class Latitude  \
0  01 ONE FAMILY DWELLINGS  1121 MADISON STREET                 40.69892   
1  01 ONE FAMILY DWELLINGS  1121 MADISON STREET                 40.69892   
2  01 ONE FAMILY DWELLINGS  1121 MADISON STREET                 40.69892   
3  01 ONE FAMILY DWELLINGS  1121 MADISON STREET                 40.69892   
4  01 ONE FAMILY DWELLINGS  1121 MADISON STREET                 40.69892   

   Building Class Longitude                     Venue  Distance  \
0                -73.909293                    Ltauha       258   
1                -73.909293  Sabor Mexicano Food Cart       217   
2                -73.909293          The Bad Old Days       273   
3                -73.909293              L'imprimerie       301   
4                -73.909293       Irving Gourmet Deli       299   

   Venue Latitude  Venue Longitude       Venue Category  
0       40.700140       -73.906691  American Restaurant  
1       40.699794       -73.911598           Food Truck  
2       40.701195       -73.908074         Cocktail Bar  
3       40.699270       -73.912834               Bakery  
4       40.697935       -73.912598       Sandwich Place

In [18]:
brooklyn_venues.groupby(by=['Address']).count().head()

Building Class Name  Building Class Latitude  \
Address                                                             
1002 HALSEY STREET                   100                      100   
101 CORNELIA STREET                  100                      100   
101 SUYDAM STREET                    100                      100   
1012 DECATUR STREET                  100                      100   
1014 DECATUR STREET                  100                      100   

                     Building Class Longitude  Venue  Distance  \
Address                                                          
1002 HALSEY STREET                        100    100       100   
101 CORNELIA STREET                       100    100       100   
101 SUYDAM STREET                         100    100       100   
1012 DECATUR STREET                       100    100       100   
1014 DECATUR STREET                       100    100       100   

                     Venue Latitude  Venue Longitude  Venue Category  
Address                                                               
1002 HALSEY STREET              100              100             100  
101 CORNELIA STREET             100              100             100  
101 SUYDAM STREET               100              100             100  
1012 DECATUR STREET             100              100             100  
1014 DECATUR STREET             100              100             100

### Analyzing Venues

Now we are going to convert the categorical data to numerical data using one hot encoding.

In [19]:
brooklyn_onehot = pd.get_dummies(brooklyn_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
brooklyn_onehot['Address'] = brooklyn_venues['Address'] 

# move neighborhood column to the first column
fixed_columns = [brooklyn_onehot.columns[brooklyn_onehot.columns=='Address'][0]] + list(brooklyn_onehot.columns[brooklyn_onehot.columns!='Address'])
brooklyn_onehot = brooklyn_onehot[fixed_columns]

brooklyn_onehot.head()

Address  African Restaurant  American Restaurant  Arcade  \
0  1121 MADISON STREET                   0                    1       0   
1  1121 MADISON STREET                   0                    0       0   
2  1121 MADISON STREET                   0                    0       0   
3  1121 MADISON STREET                   0                    0       0   
4  1121 MADISON STREET                   0                    0       0   

   Arepa Restaurant  Art Gallery  Art Museum  Arts & Crafts Store  \
0                 0            0           0                    0   
1                 0            0           0                    0   
2                 0            0           0                    0   
3                 0            0           0                    0   
4                 0            0           0                    0   

   Asian Restaurant  Athletics & Sports  BBQ Joint  Bagel Shop  Bakery  Bank  \
0                 0                   0          0           0       0     0   
1                 0                   0          0           0       0     0   
2                 0                   0          0           0       0     0   
3                 0                   0          0           0       1     0   
4                 0                   0          0           0       0     0   

   Bar  Baseball Field  Bed & Breakfast  Beer Bar  Beer Garden  Bistro  \
0    0               0                0         0            0       0   
1    0               0                0         0            0       0   
2    0               0                0         0            0       0   
3    0               0                0         0            0       0   
4    0               0                0         0            0       0   

   Bookstore  Botanical Garden  Boutique  Breakfast Spot  Brewery  \
0          0                 0         0               0        0   
1          0                 0         0               0        0   
2          0                 0         0               0        0   
3          0                 0         0               0        0   
4          0                 0         0               0        0   

   Burger Joint  Burrito Place  Bus Station  Bus Stop  Café  \
0             0              0            0         0     0   
1             0              0            0         0     0   
2             0              0            0         0     0   
3             0              0            0         0     0   
4             0              0            0         0     0   

   Cajun / Creole Restaurant  Candy Store  Caribbean Restaurant  \
0                          0            0                     0   
1                          0            0                     0   
2                          0            0                     0   
3                          0            0                     0   
4                          0            0                     0   

   Chinese Restaurant  Chocolate Shop  Climbing Gym  Clothing Store  \
0                   0               0             0               0   
1                   0               0             0               0   
2                   0               0             0               0   
3                   0               0             0               0   
4                   0               0             0               0   

   Cocktail Bar  Coffee Shop  Comedy Club  Comfort Food Restaurant  \
0             0            0            0                        0   
1             0            0            0                        0   
2             1            0            0                        0   
3             0            0            0                        0   
4             0            0            0                        0   

   Convenience Store  Cosmetics Shop  Coworking Space  Cuban Restaurant  \
0                  0               0                0                 0   
1                  0        

### Note: I am adding SALE PRICE and Years also in the following table which we will be using in the K-Means algorithm.

In [20]:
brooklyn_onehot_group = brooklyn_onehot.groupby('Address').mean().reset_index()
brooklyn_onehot_group=brooklyn_onehot_group.merge(brook_prop[['SALE PRICE','ADDRESS','Total Years']],left_on='Address',right_on='ADDRESS')
brooklyn_onehot_group.drop(columns=['ADDRESS'],inplace=True)
brooklyn_onehot_group.head()


Address  African Restaurant  American Restaurant  Arcade  \
0   1002 HALSEY STREET                0.01                 0.00     0.0   
1  101 CORNELIA STREET                0.00                 0.02     0.0   
2    101 SUYDAM STREET                0.00                 0.01     0.0   
3  1012 DECATUR STREET                0.01                 0.00     0.0   
4  1014 DECATUR STREET                0.01                 0.00     0.0   

   Arepa Restaurant  Art Gallery  Art Museum  Arts & Crafts Store  \
0              0.00         0.00         0.0                  0.0   
1              0.00         0.01         0.0                  0.0   
2              0.01         0.00         0.0                  0.0   
3              0.00         0.01         0.0                  0.0   
4              0.00         0.01         0.0                  0.0   

   Asian Restaurant  Athletics & Sports  BBQ Joint  Bagel Shop  Bakery  Bank  \
0               0.0                 0.0       0.02        0.00    0.03   0.0   
1               0.0                 0.0       0.00        0.00    0.05   0.0   
2               0.0                 0.0       0.00        0.01    0.04   0.0   
3               0.0                 0.0       0.01        0.00    0.03   0.0   
4               0.0                 0.0       0.01        0.00    0.03   0.0   

    Bar  Baseball Field  Bed & Breakfast  Beer Bar  Beer Garden  Bistro  \
0  0.07             0.0             0.01      0.01         0.00     0.0   
1  0.12             0.0             0.00      0.01         0.01     0.0   
2  0.17             0.0             0.00      0.00         0.00     0.0   
3  0.05             0.0             0.01      0.01         0.00     0.0   
4  0.05             0.0             0.01      0.01         0.00     0.0   

   Bookstore  Botanical Garden  Boutique  Breakfast Spot  Brewery  \
0       0.00               0.0      0.02             0.0     0.00   
1       0.01               0.0      0.00             0.0     0.03   
2       0.00               0.0      0.00             0.0     0.01   
3       0.00               0.0      0.02             0.0     0.00   
4       0.00               0.0      0.02             0.0     0.00   

   Burger Joint  Burrito Place  Bus Station  Bus Stop  Café  \
0          0.00            0.0         0.00       0.0  0.07   
1          0.00            0.0         0.01       0.0  0.02   
2          0.01            0.0         0.00       0.0  0.01   
3          0.00            0.0         0.00       0.0  0.08   
4          0.00            0.0         0.00       0.0  0.08   

   Cajun / Creole Restaurant  Candy Store  Caribbean Restaurant  \
0                        0.0          0.0                  0.06   
1                        0.0          0.0                  0.01   
2                        0.0          0.0                  0.01   
3                        0.0          0.0                  0.05   
4                        0.0          0.0                  0.05   

   Chinese Restaurant  Chocolate Shop  Climbing Gym  Clothing Store  \
0                0.03             0.0          0.00            0.01   
1                0.04             0.0          0.00            0.00   
2                0.02             0.0          0.01            0.00   
3                0.02             0.0          0.00            0.01   
4                0.02             0.0          0.00            0.01   

   Cocktail Bar  Coffee Shop  Comedy Club  Comfort Food Restaurant  \
0          0.01         0.02          0.0                     0.02   
1          0.02         0.05          0.0                     0.00   
2          0.03         0.05          0.0                     0.00   
3          0.02         0.05          0.0                     0.02   
4          0.02         0.05          0.0                     0.02   

   Convenience Store  Cosmetics Shop  Coworking Space  Cuban Restaurant  \
0               0.00             0.0             0.00               0.0   
1               0.01  

In [21]:
def most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

### Getting most common venues

In [22]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Address']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
brooklyn_venues_sorted = pd.DataFrame(columns=columns)
brooklyn_venues_sorted['Address'] = brooklyn_onehot_group['Address']

for ind in np.arange(brooklyn_onehot_group.shape[0]):
  brooklyn_venues_sorted.iloc[ind, 1:len(brooklyn_onehot_group.columns)-2] = most_common_venues(brooklyn_onehot_group.iloc[ind, :-2], num_top_venues)

brooklyn_venues_sorted.head()

Address 1st Most Common Venue 2nd Most Common Venue  \
0   1002 HALSEY STREET                   Bar           Pizza Place   
1  101 CORNELIA STREET                   Bar           Pizza Place   
2    101 SUYDAM STREET                   Bar    Mexican Restaurant   
3  1012 DECATUR STREET                  Café           Coffee Shop   
4  1014 DECATUR STREET                  Café           Coffee Shop   

  3rd Most Common Venue 4th Most Common Venue  \
0                  Café  Caribbean Restaurant   
1    Mexican Restaurant                Bakery   
2           Pizza Place           Coffee Shop   
3        Discount Store                   Bar   
4        Discount Store                   Bar   

             5th Most Common Venue            6th Most Common Venue  \
0  Southern / Soul Food Restaurant                    Deli / Bodega   
1                      Coffee Shop        Latin American Restaurant   
2                           Bakery                      Yoga Studio   
3             Caribbean Restaurant  Southern / Soul Food Restaurant   
4             Caribbean Restaurant  Southern / Soul Food Restaurant   

  7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
0        Discount Store        Sandwich Place                Bakery   
1         Deli / Bodega    Chinese Restaurant             Nightclub   
2          Cocktail Bar          Gourmet Shop    Italian Restaurant   
3         Deli / Bodega           Pizza Place                  Park   
4         Deli / Bodega           Pizza Place                  Park   

  10th Most Common Venue  
0     Chinese Restaurant  
1                Brewery  
2     Chinese Restaurant  
3     Mexican Restaurant  
4     Mexican Restaurant

## 3.Modeling

#### Model training 

In [23]:
kclusters = 3

brooklyn_grouped_clustering = brooklyn_onehot_group.drop('Address', 1)

from sklearn import preprocessing
X=np.asarray(brooklyn_grouped_clustering)
X=preprocessing.StandardScaler().fit_transform(X)
# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(X)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10]

array([1, 2, 0, 1, 1, 1, 1, 1, 1, 2], dtype=int32)

In [24]:
brooklyn_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

brooklyn_merged = brook_prop

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
brooklyn_merged = brooklyn_merged.join(brooklyn_venues_sorted.set_index('Address'), on='ADDRESS')

brooklyn_merged.head() # check the last columns!

BUILDING CLASS CATEGORY              ADDRESS  SALE PRICE  Total Years  \
4927  01 ONE FAMILY DWELLINGS  1121 MADISON STREET  11750000.0        162.0   
4929  01 ONE FAMILY DWELLINGS  1095 MADISON STREET   3600000.0        119.0   
4931  01 ONE FAMILY DWELLINGS  1249 MADISON STREET   2530000.0        120.0   
4933  01 ONE FAMILY DWELLINGS   1245 PUTNAM AVENUE   5000000.0        108.0   
4934  01 ONE FAMILY DWELLINGS   1241 PUTNAM AVENUE   8325000.0         98.0   

       latitude  longitude  Cluster Labels 1st Most Common Venue  \
4927  40.698920 -73.909293               2                   Bar   
4929  40.698920 -73.909293               2                   Bar   
4931  40.698920 -73.909293               2                   Bar   
4933  40.682587 -73.962611               0                   Bar   
4934  40.682587 -73.962611               0                   Bar   

     2nd Most Common Venue 3rd Most Common Venue      4th Most Common Venue  \
4927    Mexican Restaurant           Coffee Shop  Latin American Restaurant   
4929    Mexican Restaurant           Coffee Shop  Latin American Restaurant   
4931    Mexican Restaurant           Coffee Shop  Latin American Restaurant   
4933          Cocktail Bar                  Café                Pizza Place   
4934          Cocktail Bar                  Café                Pizza Place   

        5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
4927              Pizza Place                Bakery                   Gym   
4929              Pizza Place                Bakery                   Gym   
4931              Pizza Place                Bakery                   Gym   
4933  New American Restaurant                Bakery    Italian Restaurant   
4934  New American Restaurant                Bakery    Italian Restaurant   

     8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
4927               Brewery             Nightclub          Deli / Bodega  
4929               Brewery             Nightclub          Deli / Bodega  
4931               Brewery             Nightclub          Deli / Bodega  
4933             Wine Shop           Coffee Shop            Yoga Studio  
4934             Wine Shop           Coffee Shop            Yoga Studio

## Plotting the Cluster map.

In [83]:
map_clusters = folium.Map(location=[latitudes, longitudes], zoom_start=12)

# set color scheme for the clusters

rainbow = ['black','red','yellow','#ff66cc','blue']

markers_colors = []
for lat, lon, poi, price,cluster in zip(brooklyn_merged['latitude'], brooklyn_merged['longitude'], brooklyn_merged['BUILDING CLASS CATEGORY'],brooklyn_merged['SALE PRICE'] ,brooklyn_merged['Cluster Labels']):
    label = folium.Popup(str(poi) +'<br>Sale Price :'+ '${:,.2f}'.format(price) +' <br>Cluster : ' + str(cluster))
    folium.CircleMarker(
        [lat, lon],
        radius=4,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters.save('Cluster.html')

## Analyzing Clusters

#### Analyzing each cluster

### Cluster 0

In [26]:

cluster0=brooklyn_merged.loc[brooklyn_merged['Cluster Labels'] == 0, brooklyn_merged.columns[[0] + list(range(1, brooklyn_merged.shape[1]))]]
cluster0.head(10)

BUILDING CLASS CATEGORY               ADDRESS  SALE PRICE  \
4933         01 ONE FAMILY DWELLINGS    1245 PUTNAM AVENUE   5000000.0   
4934         01 ONE FAMILY DWELLINGS    1241 PUTNAM AVENUE   8325000.0   
4939         01 ONE FAMILY DWELLINGS    1361 PUTNAM AVENUE   5500000.0   
4942         02 TWO FAMILY DWELLINGS       1294 PUTNAM AVE   7400000.0   
4943         02 TWO FAMILY DWELLINGS    1308 PUTNAM AVENUE   5500000.0   
5057  10 COOPS - ELEVATOR APARTMENTS  2048 EASTERN PARKWAY    435000.0   
5062  10 COOPS - ELEVATOR APARTMENTS     101 SUYDAM STREET    985031.0   
5064  10 COOPS - ELEVATOR APARTMENTS       541 HART STREET    985000.0   
5068  10 COOPS - ELEVATOR APARTMENTS    1146 DEKALB AVENUE    945000.0   
5070  10 COOPS - ELEVATOR APARTMENTS    1376 DEKALB AVENUE   1120000.0   

      Total Years   latitude  longitude  Cluster Labels 1st Most Common Venue  \
4933        108.0  40.682587 -73.962611               0                   Bar   
4934         98.0  40.682587 -73.962611               0                   Bar   
4939        117.0  40.682587 -73.962611               0                   Bar   
4942        119.0  40.682587 -73.962611               0                   Bar   
4943        119.0  40.682587 -73.962611               0                   Bar   
5057         80.0  40.672570 -73.966080               0                   Bar   
5062         81.0  40.707332 -73.918532               0                   Bar   
5064         80.0  40.706947 -73.917740               0                   Bar   
5068         98.0  40.706546 -73.916948               0                   Bar   
5070         95.0  40.706546 -73.916948               0                   Bar   

     2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
4933          Cocktail Bar                  Café           Pizza Place   
4934          Cocktail Bar                  Café           Pizza Place   
4939          Cocktail Bar                  Café           Pizza Place   
4942          Cocktail Bar                  Café           Pizza Place   
4943          Cocktail Bar                  Café           Pizza Place   
5057                Garden          Cocktail Bar      Sushi Restaurant   
5062    Mexican Restaurant           Pizza Place           Coffee Shop   
5064    Mexican Restaurant           Pizza Place           Coffee Shop   
5068    Mexican Restaurant           Pizza Place           Coffee Shop   
5070    Mexican Restaurant           Pizza Place           Coffee Shop   

        5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
4933  New American Restaurant                Bakery    Italian Restaurant   
4934  New American Restaurant                Bakery    Italian Restaurant   
4939  New American Restaurant                Bakery    Italian Restaurant   
4942  New American Restaurant                Bakery    Italian Restaurant   
4943  New American Restaurant                Bakery    Italian Restaurant   
5057                Wine Shop           Pizza Place                 Plaza   
5062                   Bakery           Yoga Studio          Cocktail Bar   
5064                   Bakery          Cocktail Bar          Liquor Store   
5068                   Bakery          Cocktail Bar          Liquor Store   
5070                   Bakery          Cocktail Bar          Liquor Store   

     8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
4933             Wine Shop           Coffee Shop            Yoga Studio  
4934             Wine Shop           Coffee Shop            Yoga Studio  
4939             Wine Shop           Coffee Shop            Yoga Studio  
4942             Wine Shop           Coffee Shop            Yoga Studio  
4943             Wine Shop           Coffee Shop            Yoga Studio  
5057                Bakery           Coffee Shop    American Restaurant  
5062          Gourmet Shop    Italian Restaurant     Chinese Restaurant  
5064    Chinese Restaurant           Music Venue   

### Cluster 1

In [27]:
cluster1=brooklyn_merged.loc[brooklyn_merged['Cluster Labels'] == 1, brooklyn_merged.columns[[0] + list(range(1, brooklyn_merged.shape[1]))]]
cluster1.head(10)

BUILDING CLASS CATEGORY                ADDRESS  SALE PRICE  \
4951           02 TWO FAMILY DWELLINGS     997 HANCOCK STREET   6900000.0   
4952           02 TWO FAMILY DWELLINGS  1118 JEFFERSON AVENUE   8100000.0   
4954           02 TWO FAMILY DWELLINGS    1055 HANCOCK STREET   6500000.0   
4957         03 THREE FAMILY DWELLINGS  1328 JEFFERSON AVENUE   6000000.0   
4985    07 RENTALS - WALKUP APARTMENTS     1183 HALSEY STREET   2300000.0   
4987  08 RENTALS - ELEVATOR APARTMENTS     1002 HALSEY STREET  14920000.0   
4989      09 COOPS - WALKUP APARTMENTS     1238 HALSEY STREET   3875000.0   
5015      09 COOPS - WALKUP APARTMENTS    1012 DECATUR STREET    618000.0   
5016      09 COOPS - WALKUP APARTMENTS    1016 DECATUR STREET    583000.0   
5017      09 COOPS - WALKUP APARTMENTS    1014 DECATUR STREET   2400000.0   

      Total Years   latitude  longitude  Cluster Labels 1st Most Common Venue  \
4951        117.0  40.684228 -73.935301               1                   Bar   
4952        117.0  40.684627 -73.938366               1                   Bar   
4954        117.0  40.684228 -73.935301               1                   Bar   
4957        179.0  40.684627 -73.938366               1                   Bar   
4985        169.0  40.683840 -73.932231               1                   Bar   
4987        119.0  40.683840 -73.932231               1                   Bar   
4989        157.0  40.683840 -73.932231               1                   Bar   
5015        118.0  40.681970 -73.928867               1                  Café   
5016        119.0  40.681970 -73.928867               1                  Café   
5017        118.0  40.681970 -73.928867               1                  Café   

     2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
4951                  Café           Pizza Place  Caribbean Restaurant   
4952           Coffee Shop                  Café           Pizza Place   
4954                  Café           Pizza Place  Caribbean Restaurant   
4957           Coffee Shop                  Café           Pizza Place   
4985           Pizza Place                  Café  Caribbean Restaurant   
4987           Pizza Place                  Café  Caribbean Restaurant   
4989           Pizza Place                  Café  Caribbean Restaurant   
5015           Coffee Shop        Discount Store                   Bar   
5016           Coffee Shop        Discount Store                   Bar   
5017           Coffee Shop        Discount Store                   Bar   

                5th Most Common Venue            6th Most Common Venue  \
4951                      Coffee Shop                    Deli / Bodega   
4952                        Wine Shop                   Discount Store   
4954                      Coffee Shop                    Deli / Bodega   
4957                        Wine Shop                   Discount Store   
4985  Southern / Soul Food Restaurant                    Deli / Bodega   
4987  Southern / Soul Food Restaurant                    Deli / Bodega   
4989  Southern / Soul Food Restaurant                    Deli / Bodega   
5015             Caribbean Restaurant  Southern / Soul Food Restaurant   
5016             Caribbean Restaurant  Southern / Soul Food Restaurant   
5017             Caribbean Restaurant  Southern / Soul Food Restaurant   

     7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
4951    Chinese Restaurant        Discount Store             Wine Shop   
4952  Caribbean Restaurant         Deli / Bodega        Sandwich Place   
4954    Chinese Restaurant        Discount Store             Wine Shop   
4957  Caribbean Restaurant         Deli / Bodega        Sandwich Place   
4985        Discount Store        Sandwich Place                Bakery   
4987        Discount Store        Sandwich Place                Bakery   
4989        Discount Store        Sandwich Place                Bakery   
5015         Deli / Bodega           Pizza Place     

### Cluster 2

In [28]:
cluster2=brooklyn_merged.loc[brooklyn_merged['Cluster Labels'] == 2, brooklyn_merged.columns[[0] + list(range(1, brooklyn_merged.shape[1]))]]
cluster2.head(10)

BUILDING CLASS CATEGORY               ADDRESS  SALE PRICE  \
4927         01 ONE FAMILY DWELLINGS   1121 MADISON STREET  11750000.0   
4929         01 ONE FAMILY DWELLINGS   1095 MADISON STREET   3600000.0   
4931         01 ONE FAMILY DWELLINGS   1249 MADISON STREET   2530000.0   
4935         01 ONE FAMILY DWELLINGS   1168 MADISON STREET   5350000.0   
4940         02 TWO FAMILY DWELLINGS    77 CORNELIA STREET  12000000.0   
4944         02 TWO FAMILY DWELLINGS          207 CORNELIA   4300000.0   
4945         02 TWO FAMILY DWELLINGS   339 CORNELIA STREET   5700000.0   
4966          06 TAX CLASS 1 - OTHER  149 WEIRFIELD STREET   2500000.0   
4967  07 RENTALS - WALKUP APARTMENTS  137 WEIRFIELD STREET  15500000.0   
4971  07 RENTALS - WALKUP APARTMENTS  205 WEIRFIELD STREET   3200000.0   

      Total Years   latitude  longitude  Cluster Labels 1st Most Common Venue  \
4927        162.0  40.698920 -73.909293               2                   Bar   
4929        119.0  40.698920 -73.909293               2                   Bar   
4931        120.0  40.698920 -73.909293               2                   Bar   
4935        119.0  40.698920 -73.909293               2                   Bar   
4940        119.0  40.697825 -73.908026               2                   Bar   
4944        178.0  40.697825 -73.908026               2                   Bar   
4945        104.0  40.695843 -73.910250               2                   Bar   
4966         32.0  40.689927 -73.912550               2                  Café   
4967        178.0  40.689679 -73.912819               2           Pizza Place   
4971        177.0  40.691315 -73.911186               2           Coffee Shop   

     2nd Most Common Venue 3rd Most Common Venue      4th Most Common Venue  \
4927    Mexican Restaurant           Coffee Shop  Latin American Restaurant   
4929    Mexican Restaurant           Coffee Shop  Latin American Restaurant   
4931    Mexican Restaurant           Coffee Shop  Latin American Restaurant   
4935    Mexican Restaurant           Coffee Shop  Latin American Restaurant   
4940           Pizza Place    Mexican Restaurant                     Bakery   
4944           Pizza Place    Mexican Restaurant                     Bakery   
4945           Pizza Place           Coffee Shop              Deli / Bodega   
4966           Pizza Place         Deli / Bodega                Coffee Shop   
4967                  Café        Sandwich Place              Deli / Bodega   
4971           Pizza Place                  Café         Chinese Restaurant   

     5th Most Common Venue      6th Most Common Venue  \
4927           Pizza Place                     Bakery   
4929           Pizza Place                     Bakery   
4931           Pizza Place                     Bakery   
4935           Pizza Place                     Bakery   
4940           Coffee Shop  Latin American Restaurant   
4944           Coffee Shop  Latin American Restaurant   
4945    Mexican Restaurant                     Bakery   
4966        Sandwich Place              Grocery Store   
4967    Chinese Restaurant        Fried Chicken Joint   
4971                   Bar             Discount Store   

          7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
4927                        Gym               Brewery             Nightclub   
4929                        Gym               Brewery             Nightclub   
4931                        Gym               Brewery             Nightclub   
4935                        Gym               Brewery             Nightclub   
4940              Deli / Bodega    Chinese Restaurant             Nightclub   
4944              Deli / Bodega    Chinese Restaurant             Nightclub   
4945  Latin American Restaurant                   Gym    Chinese Restaurant   
4966                 Restaurant               Brewery    Chinese Restaurant   
4967          Indian Restaurant         Grocery Store           Supermarket   
4971              

## Result and Discussion

According to the K-Means clustering model we can see that there are 3 clusters. Most of the property comes under cluster 0 and cluster 2. Now lets see what comes under cluster 0 and cluster 2.

Custer 0: In this cluster most of the common venues are bars, coffee shops, wine shops and yoga studio. There are few restaurants also such as Italian Restaurants. So we can say that a person who likes to do party and love wine will love this property.

Cluster 2: In this cluster most of the venues are restaurants such as Pizza place, Chinese Restaurant, Mexican Restaurant, Sandwich Place, Latin American Restaurant, Fried Chicken Joint and many more. There few Bar and Nighclub also. This cluster also have Supermarket and Grocery Store. This kind of property is very much likeable because every think is available in this area.

You can see below the maximun price of the Cluster 0, 1 and 2. Cluster 2 have the most expensive property.

In [47]:
print('Maximum price of Property in Cluster 0 is : {} million'.format(cluster0['SALE PRICE'].max()/1000000))
print('Maximum price of Property in Cluster 1 is : {} million'.format(cluster1['SALE PRICE'].max()/1000000))
print('Maximum price of Property in Cluster 2 is : {} million'.format(cluster2['SALE PRICE'].max()/1000000))

Maximum price of Property in Cluster 0 is : 8.325 million
Maximum price of Property in Cluster 1 is : 14.92 million
Maximum price of Property in Cluster 2 is : 21.9 million


## Conclusion

Final Conclusion of this report is that most of the property comes under the cluster 0 and 2 but most favourable property will be in cluster 2 because it contains all the basic amenities and essential facilities. Cluster 2 has better property as compared to cluster 0. 

If we also see the price range of the property in both the clusters, Cluster 0 have a price range of 0.151 million dollar to  8.325 Miilion dollar whereas Cluster 2 have price range of 0.32 million dollar to 21.9 million doolar. 